include/xgboost/gbm.h

```diff
@@ -14,17 +14,22 @@
 #include <string>
 #include <functional>
 #include <memory>
-#include "./base.h"
-#include "./data.h"
-#include "./objective.h"
-#include "./feature_map.h"
+
+#include <xgboost/base.h>
+#include <xgboost/data.h>
+#include <xgboost/objective.h>
+#include <xgboost/feature_map.h>
+#include <xgboost/generic_parameters.h>
 #include "../../src/common/host_device_vector.h"
 
 namespace xgboost {
 /*!
  * \brief interface of gradient boosting model.
  */
 class GradientBooster {
+ protected:
+  LearnerTrainParam const* learner_param_;
```

Comment:
No copying is because `Learner`(booster) can change its parameters.  Users can try both GPU and CPU in a single script or REPL.

Smart pointers imply ownership sharing/transferring, but I don't want other components other than `Learner`(booster) to change the parameters, hence a raw pointer to constant.

Reply:
It looks like the same object can be pointed to by `std::shared_ptr<T>` and `std::shared_ptr<const T>`: constructor (9) (the templated version) in https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr appears to support this case.

Disclaimer: I'm not a C++ expert, so I may be wrong here.

Reply:
`std::shared_ptr<T>` indeed does not copy the underlying object. However, it performs reference counting, which involves atomic adds and subtracts, and thus adds overhead.

Reply:
Yes, it uses atomics, but only during initialization. Dereferencing and member access don't use them.

I'll leave the decision up to you.